### PR TITLE
Context `merge` function

### LINKED ISSUE #1013: Add a merge API to `cedar_policy::Context`
### Category

User level API features/changes

### Describe the feature you'd like to request

As a developer working on translation of user-defined rust types to `cedar_policy::Context`, I find that I often want to reach for a `merge` method to merge two contexts `ctx1.merge(ctx2)`. A use-case for this is when translating nested structs of attributes into a nested `Context`.

Consider a top-level aggregate of multiple user-defined Contexts.
```rust
#[derive(Builder, Debug, Clone, PartialEq, Eq)]
#[builder(pattern = "owned")]
pub struct MyContext {
    pub foo: FooContext,
    pub bar: BarContext,
}
```

Where sub-contexts may have other nested user-defined Contexts that I need to translate into a `cedar_policy::Context`, are defined as:

```rust
// Contains other Contexts I need to translate, recursively.
pub struct FooContext {
    pub ctx_a: Option<FooContextA>,
    pub ctx_b: Option<FooContextB>,
}

// A leaf Context.
pub struct BarContext {
    pub key: String,
    pub value: String,
}
```
A merge API could help write code like this to stitch together sub-context translation results:

```rust
impl MyContext {
    pub fn try_into_cedar_context(self) -> Result<cedar_policy::Context, MyContextToCedarContextError> {
            let foo_ctx = self.foo.try_into_cedar_context()?;
            let bar_ctx = self.bar.try_into_cedar_context()?;
            Ok(foo_ctx.merge(bar_ctx));
    }
}

// Example impl for a non-top-level context.
impl FooContext {
    // It would be good to run a fold/collect to reduce over the contexts. Rough idea, may not compile.
    pub fn try_into_cedar_context(self) -> Result<cedar_policy::Context, MyContextToCedarContextError> {
          Ok(vec![self.ctx_a, self.ctx_b].iter()
                    .fold(Context::new(), |mut acc, ctx| { 
                         acc.merge(ctx);
                         acc
                  })?);
    }
}
```

### Describe alternatives you've considered

Aside from the `merge` proposal another good alternative could be implementing the `extend` trait like `HashMap` does, https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.extend.

The workaround that I ended up doing was to implement non-top-level translations to a `HashMap<String, RestrictedExpression>`. Then I can build the top-level context by merging the restricted expression maps and passing that to `Context::from_pairs(result)`. 

```rust
// Translation of a leaf context.
impl BarContext {
    pub fn try_into_cedar_context(self) -> Result<HashMap<String, RestrictedExpression, MyContextToCedarContextError> {
          Ok(HashMap::from([
               ("key".to_string(), RestrictedExpression::new_string(self.key)),
               ("value".to_string(), RestrictedExpression::new_string(self.value)),
           ])?);
    }
}
```

At the top-level I converted the sub-contexts to a `RestrictedExpression` record, and then used `Context::from_pairs`:

```rust
impl MyContext {
    pub fn try_into_cedar_context(self) -> Result<cedar_policy::Context, MyContextToCedarContextError> {
         let foo_expr_map = self.foo.try_into_cedar_context()?;
         let bar_expr_map = self.bar.try_into_cedar_context()?;
         Ok(Context::from_pairs([
             ("foo".to_string(), RestrictedExpression::new_record(foo_expr_map.into_iter())?),
             ("bar".to_string(), RestrictedExpression::new_record(bar_expr_map.into_iter())?)
         ]))
    }
}
```

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by khieta**
## Description of changes

Adds two new methods to the public API:
* `Context::merge` as suggested in #1013 
* `Context::values` which gets the key/value pairs out of the context. This function seemed like the most natural way to implement `merge`, and could be useful in general.

Makes a breaking error API change (because I can't help myself):
* Renamed `DuplicateKeyInRecordLiteralError` to `DuplicateKeyError` and added a "context" field to allow more precise error messages. The reason for this change is just that I didn't like the error message "duplicate key _ in a record literal" for the `Context` functions.

## Issue #, if available

Resolves #1013

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

**Review by shaobo-he-aws — APPROVED** [review]

File: cedar-policy-core/src/ast/expr.rs
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/expr.rs:1160]: probably rename the variant here too while we're at it

File: cedar-policy-core/src/ast/expr.rs
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/expr.rs:1190]: I'm not sure if this is general idiomaticity guideline or not, but I would expect `set_` to be a method that takes `&mut Self`, and `with_` to be a method that takes `self` and returns `Self`, like `Value::with_maybe_source_loc()` in Core.  Take or leave this advice

File: cedar-policy-core/src/ast/request.rs
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/request.rs:323]: Should we call this `.into_iter()` instead?  Maybe even have it be an implementation of the `IntoIterator` trait?

File: cedar-policy/src/api.rs
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:3405]: might be useful to have a `ast::RestrictedExpr::from(val)` function.  From this vantage point it's a little harder to see why `new_unchecked` is safe here

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-core/src/ast/expr.rs
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/ast/expr.rs:1190]: I wouldn't make this `pub`. `pub(crate)` probably?

File: cedar-policy-core/src/ast/request.rs
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/ast/request.rs:333]: How is the `Unknown` case handled now? Kind of worried it might hit the `panic!("internal invariant violation: expected a record, got {kind:?}"),` case

Comment (khieta) [reply cedar-policy-core/src/ast/request.rs:340]: Great question! I believe the `Context::from_expr` function is used on all paths for building a context, and this function must return a record value or record residual due to the handing of records in `RestrictedEvaluator::partial_interpret_internal`. I updated the comment about the `ContextRecord` invariant to reflect this, but would appreciate if you would double check my reasoning.

File: cedar-policy-core/src/ast/request.rs
Comment (khieta) [reply cedar-policy-core/src/ast/request.rs:327]: I'll try this out tomorrow

File: cedar-policy-core/src/ast/request.rs
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/ast/request.rs:340]: Looked at the code closer and I agree that this match is safe given the invariant. 

#1026 introduced `from_partial_value_unchecked` which makes the invariant depend on those call sites. I'd be happier if that didn't need the `_unchecked` constructor, but I am convinced that the invariant is maintained. We already have #541 for future refactoring.

File: cedar-policy-core/src/entities/json/err.rs
Comment (aaronjeline) [inline cedar-policy-core/src/entities/json/err.rs:86]: This is a nicer name

**Review by aaronjeline — APPROVED** [review]

File: cedar-policy-core/src/ast/request.rs
Comment (khieta) [reply cedar-policy-core/src/ast/request.rs:340]: Latest version adds additional comments to `from_partial_value_unchecked` to clarify 🙂

File: cedar-policy-core/src/ast/request.rs
Comment (khieta) [reply cedar-policy-core/src/ast/request.rs]: Latest version adds an implementation of `IntoIterator`. I don't think we can provide an `.iter()` function that doesn't perform cloning atm. I think the best solution would be to address #541 by storing the context as a hash map instead of a record expression.

File: cedar-policy-core/src/ast/request.rs
Comment (shaobo-he-aws) [reply cedar-policy-core/src/ast/request.rs:340]: Would it make sense to implement #541 in this PR if it's convenient?

File: cedar-policy/src/api.rs
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:3416]: maybe a short comment justifying the `_unchecked`, like you have elsewhere in this PR?

**Review by cdisselkoen — APPROVED** [review]
Looks great

File: cedar-policy-core/src/ast/request.rs
Comment (cdisselkoen) [reply cedar-policy-core/src/ast/request.rs]: Nicely done :)

Want to add a comment to #541 with your suggestion for implementation?

File: cedar-policy/src/api.rs
Comment (shaobo-he-aws) [inline cedar-policy/src/api.rs:3389]: Sorry for being late. I was wondering if we can generalize the argument to `impl Iterator<SmolStr, ...>`, since `Context` implements the conversion.

File: cedar-policy-core/src/ast/request.rs
Comment (john-h-kastner-aws) [reply cedar-policy-core/src/ast/request.rs:340]: I expect that change to be a bit annoying, so I'd rather see it in a follow up PR. Up to Kesha if wants to do it ofc.

**Review by john-h-kastner-aws — APPROVED** [review]
looks good!

File: cedar-policy-core/src/ast/request.rs
Comment (khieta) [reply cedar-policy-core/src/ast/request.rs:340]: Yeah, that will be a somewhat involved change, so I'll leave it for future work.

**Comment by khieta** [comment]
I believe I've addressed all comments now. Will wait until tomorrow to merge because I will need to make some downstream fixes (caused by the change `Context::iter` -> `Context::into_iter`)

**Comment by khieta** [comment]
Fix for DRT failure is here: https://github.com/cedar-policy/cedar-spec/pull/376. Need a re-approval before I can merge this PR.

**Pull request merged**
